src/main/java/com/exactpro/th2/act/BiDirectionalMessageReceiver.java

```diff
@@ -0,0 +1,190 @@
+        START, OUTGOING_MATCHED, INCOMING_MATCHED;
+    }
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(BiDirectionalMessageReceiver.class);
+
+    private final SubscriptionManager subscriptionManager;
+    private final CheckRule outgoingRule;
+    private final Function<Message, CheckRule> incomingRuleSupplier;
+
+    private final Queue<MessageBatch> incomingBuffer = new LinkedList<>();
+    private final MessageListener<MessageBatch> incomingListener = this::processIncomingMessages;
+    private final MessageListener<MessageBatch> outgoingListener = this::processOutgoingMessages;
+    private final AtomicReference<CheckRule> incomingRule = new AtomicReference<>();
+
+    private volatile State state = State.START;
+
+    public BiDirectionalMessageReceiver(
+            SubscriptionManager subscriptionManager,
+            ResponseMonitor monitor,
+            CheckRule outgoingRule,
+            Function<Message, CheckRule> incomingRuleSupplier
+    ) {
+        super(monitor);
+        this.subscriptionManager = Objects.requireNonNull(subscriptionManager, "'Subscription manager' parameter");
+        this.outgoingRule = Objects.requireNonNull(outgoingRule, "'Outgoing rule' parameter");
+        this.incomingRuleSupplier = Objects.requireNonNull(incomingRuleSupplier, "'Incoming rule supplier' parameter");
+
+        subscriptionManager.register(Direction.FIRST, incomingListener);
+        subscriptionManager.register(Direction.SECOND, outgoingListener);
+    }
+
+    @Override
+    @Nullable
+    public Message getResponseMessage() {
+        CheckRule rule = incomingRule.get();
+        return rule == null ? null : rule.getResponse();
+    }
+
+    @Override
+    public Collection<MessageID> processedMessageIDs() {
+        CheckRule incoming = incomingRule.get();
+        if (incoming == null || incoming.processedIDs().isEmpty()) {
+            return outgoingRule.processedIDs();
+        }
+        var messageIDS = new ArrayList<MessageID>(outgoingRule.processedIDs().size() + incoming.processedIDs().size());
+        messageIDS.addAll(outgoingRule.processedIDs());
+        messageIDS.addAll(incoming.processedIDs());
+        return messageIDS;
+    }
+
+    @Override
+    public void close() {
+        subscriptionManager.unregister(Direction.FIRST, incomingListener);
+        subscriptionManager.unregister(Direction.SECOND, outgoingListener);
+    }
+
+    private void processOutgoingMessages(String consumingTag, MessageBatch batch) {
+        State current = state;
+        if (current == State.OUTGOING_MATCHED || current == State.INCOMING_MATCHED) {
+            // already has found everything for outgoing messages
+            return;
+        }
+        if (anyMatches(batch, outgoingRule)) {
+            Message response = outgoingRule.getResponse();
+            if (response == null) {
+                throw new IllegalStateException("Rules has found match in the batch but response is 'null'");
+            }
```

Comment:
Maybe redundant, this is double-checking for the rule logic

Reply:
Yes, it is. We prevent from transitioning to the next state because it will be an invalid operation and we will get multiple exceptions later